File: src/main/python/kthSmallestElementInASortedMatrix/KthSmallestElementInASortedMatrix.py
class Solution:
    def kthSmallest(self, matrix: list[list[int]], k: int) -> int:
        n = len(matrix)
        low = matrix[0][0]
        high = matrix[-1][-1]
        while low < high:
            mid = low + (high - low) // 2
            count = 0
            c = n - 1
            for r in range(n):
                while c >= 0 and matrix[r][c] > mid:
                    c -= 1
                count += c + 1
            if count < k:
                low = mid + 1
            else:
                high = mid
        return low

    def test(self):
        tests = [
            [[[1, 5, 9], [10, 11, 13], [12, 13, 15]], 8, 13],
            [[[-5]], 1, -5],
        ]
        for test in tests:
            matrix, k, expect_result = test
            result = self.kthSmallest(matrix, k)
            print(result)
            print(result == expect_result)


if __name__ == "__main__":
    Solution().test()

File: src/main/kotlin/me/bossm0n5t3r/leetcode/kthSmallestElementInASortedMatrix/KthSmallestElementInASortedMatrix.kt
package me.bossm0n5t3r.leetcode.kthSmallestElementInASortedMatrix

class KthSmallestElementInASortedMatrix {
    class Solution {
        fun kthSmallest(matrix: Array<IntArray>, k: Int): Int {
            return matrix.flatMap { it.toList() }.sorted()[k - 1]
        }

        fun kthSmallestByUsingBiSect(matrix: Array<IntArray>, k: Int): Int {
            return biSelect(matrix, k)
        }

        private fun biSelect(matrix: Array<IntArray>, k: Int): Int {
            val n = matrix.size
            var low = matrix[0][0]
            var high = matrix[n - 1][n - 1] + 1
            while (low < high) {
                //            println("low: $low, high: $high")
                val mid = low + (high - low) / 2
                var count = 0
                var j = n - 1
                (0 until n).forEach { i ->
                    while (j >= 0 && matrix[i][j] > mid) j--
                    count += (j + 1)
                }
                if (count < k) {
                    low = mid + 1
                } else {
                    high = mid
                }
            }
            return low
        }
    }
}
